app/models/work.rb

```diff
@@ -819,65 +806,38 @@ def has_required_tags?
-      if !self.filters.include?(filter)
-        if meta
-          self.filter_taggings.create(filter_id: filter.id, inherited: true)
-        else
-          self.filters << filter
-        end
-        filter.reset_filter_count
-      elsif !meta
-        ft = self.filter_taggings.where(["filter_id = ?", filter.id]).first
-        ft.update_attribute(:inherited, false)
-      end
-      IndexQueue.enqueue_ids(Series, series.pluck(:id), :main)
-    end
-  end
+    # Only works are included in the filter count, so if a work's
+    # filter-taggings change, the FilterCount probably needs updating.
+    FilterCount.enqueue_filters(filter_taggings.map(&:filter_id))
 
-  # Removes filter_tagging relationship unless the work is tagged with more than
-  # one synonymous tags. Also updates the series index because series inherit
-  # tags from works
-  def remove_filter_tagging(tag)
-    filter = tag.filter
-    if filter
-      filters_to_remove = [filter] + filter.meta_tags
-      filters_to_remove.each do |filter_to_remove|
-        if self.filters.include?(filter_to_remove)
-          all_sub_tags = filter_to_remove.sub_tags + [filter_to_remove]
-          sub_mergers = all_sub_tags.empty? ? [] : all_sub_tags.collect(&:mergers).flatten.compact
-          all_tags_with_filter_to_remove_as_meta = all_sub_tags + sub_mergers
-          remaining_tags = self.tags - [tag]
-          if (remaining_tags & all_tags_with_filter_to_remove_as_meta).empty? # none of the remaining tags need filter_to_remove
-            self.filter_taggings.where(filter_id: filter_to_remove.id).destroy_all
-            filter_to_remove.reset_filter_count
-            filter_to_remove.update_works_index_timestamp!
-          else # we should keep filter_to_remove, but check if inheritence needs to be updated
-            direct_tags_for_filter_to_remove = filter_to_remove.mergers + [filter_to_remove]
-            if (remaining_tags & direct_tags_for_filter_to_remove).empty? # not tagged with filter or mergers directly
-              ft = self.filter_taggings.where(["filter_id = ?", filter_to_remove.id]).first
-              ft.update_attribute(:inherited, true)
-            end
-          end
-        end
-      end
-      IndexQueue.enqueue_ids(Series, series.pluck(:id), :main)
-    end
-  end
+    # From here, we only want to update works whose filter_taggings have
+    # actually changed.
+    changed_ids = filter_taggings.map(&:filterable_id)
+    return unless changed_ids.present?
+
+    # Reindex any series associated with works whose filters have changed.
+    series_ids = SerialWork.where(work_id: changed_ids).pluck(:series_id)
+    IndexQueue.enqueue_ids(Series, series_ids, queue)
 
-  # Must be called before save
-  def visibility_changed?
-    self.posted_changed? || self.restricted_changed? || self.hidden_by_admin_changed?
+    # Reindex any pseuds associated with works whose filters have changed.
+    pseud_ids = Creatorship.where(creation_id: changed_ids,
```

Comment:
Should this be limited to approved creatorships to avoid unnecessary reindexing? (I don't imagine we'll pick up many extras even if we don't limit it to approved, though.)